File: ios/RealtimeAudioRecorderModule.swift
import AVFoundation
import ExpoModulesCore

public class RealtimeAudioRecorderModule: Module, RealtimeAudioRecorderDelegate {
  var hasListeners: Bool = false
  
  public func definition() -> ModuleDefinition {
    Name("RealtimeAudioRecorder")
    
    Events("onAudioCaptured", "onCaptureComplete")
    
    OnStartObserving {
      hasListeners = true
    }
    
    OnStopObserving {
      hasListeners = false
    }
    
    Class(RealtimeAudioRecorder.self) {
      Constructor { (audioFormat: AudioFormatSettings, echoCancellationEnabled: Bool) -> RealtimeAudioRecorder in
        let recorder: RealtimeAudioRecorder = RealtimeAudioRecorder(sampleRate: audioFormat.sampleRate,
                                                                    channelCount: audioFormat.channelCount,
                                                                    audioFormat: getCommonFormat(audioFormat.encoding))!
        recorder.echoCancellationEnabled = echoCancellationEnabled
        recorder.delegate = self
        return recorder
      }
      
      AsyncFunction("startRecording") { (recorder: RealtimeAudioRecorder) in
        do {
          try recorder.startRecording()
        } catch {
          print("Error starting recording: \(error.localizedDescription)")
        }
      }
      
      AsyncFunction("stopRecording") { (recorder: RealtimeAudioRecorder) in
        recorder.stopRecording()
      }
    }
    
    View(RealtimeAudioRecorderView.self) {
      Events("onAudioCaptured", "onCaptureComplete")
      
      Prop("waveformColor") { (
        view: RealtimeAudioRecorderView,
        hexColor: UIColor
      ) in
        view.setWaveformColor(hexColor)
      }
      
      Prop("audioFormat") { (
        view: RealtimeAudioRecorderView,
        format: AudioFormatSettings
      ) in
        view.setAudioFormat(
          sampleRate: format.sampleRate,
          commonFormat: self.getCommonFormat(format.encoding),
          channels: format.channelCount
        )
      }
      
      Prop("visualizer") { (
        view: RealtimeAudioRecorderView,
        visualizer: String
      ) in
        if visualizer == "linearWaveform" {
          view.setVisualization(LinearWaveformVisualizer())
        } else if visualizer == "circularWaveform" {
          view.setVisualization(CircularWaveformVisualizer())
        } else if visualizer == "tripleCircle" {
          view.setVisualization(TripleCircleVisualizer())
        }
      }
      
      Prop("echoCancellationEnabled") { (
        view: RealtimeAudioRecorderView,
        enabled: Bool
      ) in
        view.setEchoCancellationEnabled(enabled)
      }
      
      AsyncFunction("startRecording") { (view: RealtimeAudioRecorderView) in
        view.startRecording()
      }
      
      AsyncFunction("stopRecording") { (view: RealtimeAudioRecorderView) in
        view.stopRecording()
      }
    }
  }
  
  func base64BufferReady(_ base64Audio: String) {
    if (hasListeners) {
      let event = ["audioBuffer": base64Audio]
      sendEvent("onAudioCaptured", event)
    }
  }
  
  func bufferCaptured(_ buffer: AVAudioPCMBuffer) {
  }
  
  func audioRecorderDidFinishRecording() {
    if (hasListeners) {
      sendEvent("onCaptureComplete")
    }
  }
  
  private func getCommonFormat(_ encoding: AudioEncoding) -> AVAudioCommonFormat {
    switch encoding {
    case .pcm16bitInteger:
      return .pcmFormatInt16
    case .pcm32bitInteger:
      return .pcmFormatInt32
    case .pcm32bitFloat:
      return .pcmFormatFloat32
    case .pcm64bitFloat:
      return .pcmFormatFloat64
    }
  }
  
}

File: android/src/main/java/com/paulingalls/realtimeaudio/RealtimeAudioRecorderModule.kt
package com.paulingalls.realtimeaudio

import AudioFormatSettings
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import getAndroidColor
import mapAudioEncodingToFormat
import mapChannelCountToInputFormat

class RealtimeAudioRecorderModule : Module() {
    var hasListeners = false

    override fun definition() = ModuleDefinition {
        Name("RealtimeAudioRecorder")
        Events("onAudioCaptured", "onCaptureComplete")

        OnStartObserving {
            hasListeners = true
        }

        OnStopObserving {
            hasListeners = false
        }

        Class("RealtimeAudioRecorder", RealtimeAudioRecorder::class) {
            Constructor { format: AudioFormatSettings ->
                return@Constructor RealtimeAudioRecorder(
                    format.sampleRate,
                    mapChannelCountToInputFormat(format.channelCount),
                    mapAudioEncodingToFormat(format.encoding)
                ).apply {
                    delegate = RealtimeEventDelegate(this@RealtimeAudioRecorderModule)
                }
            }

            AsyncFunction("startRecording") { recorder: RealtimeAudioRecorder ->
                recorder.startRecording()
            }
            AsyncFunction("stopRecording") { recorder: RealtimeAudioRecorder ->
                recorder.stopRecording()
            }
        }

        View(RealtimeAudioRecorderView::class) {
            Events("onAudioCaptured", "onCaptureComplete")

            Prop("waveformColor") { view: RealtimeAudioRecorderView, hexColor: String ->
                view.setVisualizationColor(getAndroidColor(hexColor))
            }

            Prop("audioFormat") { view: RealtimeAudioRecorderView,
                                  format: AudioFormatSettings ->
                view.setAudioFormat(
                    format.sampleRate,
                    mapChannelCountToInputFormat(format.channelCount),
                    mapAudioEncodingToFormat(format.encoding)
                )
            }

            AsyncFunction("startRecording") { view: RealtimeAudioRecorderView ->
                view.startRecording()
            }

            AsyncFunction("stopRecording") { view: RealtimeAudioRecorderView ->
                view.stopRecording()
            }
        }
    }

    class RealtimeEventDelegate(
        private val module: RealtimeAudioRecorderModule
    ) : RealtimeAudioBufferDelegate {
        override fun audioStringReady(base64Audio: String) {
            if (module.hasListeners) {
                module.sendEvent(
                    "onAudioCaptured", mapOf(
                        "audioBuffer" to base64Audio
                    )
                )
            }
        }

        override fun bufferReady(buffer: ByteArray) {
        }

        override fun captureComplete() {
            if (module.hasListeners) {
                module.sendEvent("onCaptureComplete", mapOf())
            }
        }
    }
}